LocalAgent exception spam during test executor shutdown

@frsyuki  Didn't we see/fix this earlier?

```
...
Gradle Test Executor 4 finished executing tests.
2016-06-30 03:08:57 +0000 [ERROR] (local-agent-0) io.digdag.core.agent.LocalAgent: Uncaught exception
org.skife.jdbi.v2.exceptions.UnableToCreateStatementException: org.h2.jdbc.JdbcSQLException: Table "QUEUED_SHARED_TASK_LOCKS" not found; SQL statement:
with recursive t (queue_id) as ((select queue_id from queued_shared_task_locks where hold_expire_time is null order by queue_id limit 1) union all select (select queue_id from queued_shared_task_locks where hold_expire_time is null and queue_id > t.queue_id order by queue_id limit 1) from t where t.queue_id is not null) select queue_id as id from t where queue_id is not null [42102-191] [statement:"with recursive t (queue_id) as ((select queue_id from queued_shared_task_locks where hold_expire_time is null order by queue_id limit 1) union all select (select queue_id from queued_shared_task_locks where hold_expire_time is null and queue_id > t.queue_id order by queue_id limit 1) from t where t.queue_id is not null) select queue_id as id from t where queue_id is not null", located:"with recursive t (queue_id) as ((select queue_id from queued_shared_task_locks where hold_expire_time is null order by queue_id limit 1) union all select (select queue_id from queued_shared_task_locks where hold_expire_time is null and queue_id > t.queue_id order by queue_id limit 1) from t where t.queue_id is not null) select queue_id as id from t where queue_id is not null", rewritten:"with recursive t (queue_id) as ((select queue_id from queued_shared_task_locks where hold_expire_time is null order by queue_id limit 1) union all select (select queue_id from queued_shared_task_locks where hold_expire_time is null and queue_id > t.queue_id order by queue_id limit 1) from t where t.queue_id is not null) select queue_id as id from t where queue_id is not null", arguments:{ positional:{}, named:{}, finder:[]}]
	at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1306)
	at org.skife.jdbi.v2.Query.fold(Query.java:173)
	at org.skife.jdbi.v2.Query.list(Query.java:82)
	at org.skife.jdbi.v2.Query.list(Query.java:75)
	at io.digdag.core.database.DatabaseTaskQueueStore.lambda$lockSharedTasks$4(DatabaseTaskQueueStore.java:195)
	at io.digdag.core.database.BasicDatabaseStoreManager.lambda$null$2(BasicDatabaseStoreManager.java:210)
	at org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:183)
	at org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:330)
	at io.digdag.core.database.BasicDatabaseStoreManager.lambda$transaction$3(BasicDatabaseStoreManager.java:210)
	at io.digdag.util.RetryExecutor.run(RetryExecutor.java:158)
	at io.digdag.util.RetryExecutor.runInterruptible(RetryExecutor.java:117)
	at io.digdag.core.database.BasicDatabaseStoreManager.transaction(BasicDatabaseStoreManager.java:206)
	at io.digdag.core.database.DatabaseTaskQueueStore.lockSharedTasks(DatabaseTaskQueueStore.java:176)
	at io.digdag.core.database.DatabaseTaskQueueFactory$DatabaseTaskQueue.lockSharedTasks(DatabaseTaskQueueFactory.java:87)
	at io.digdag.core.agent.LocalAgent.run(LocalAgent.java:66)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.h2.jdbc.JdbcSQLException: Table "QUEUED_SHARED_TASK_LOCKS" not found; SQL statement:
with recursive t (queue_id) as ((select queue_id from queued_shared_task_locks where hold_expire_time is null order by queue_id limit 1) union all select (select queue_id from queued_shared_task_locks where hold_expire_time is null and queue_id > t.queue_id order by queue_id limit 1) from t where t.queue_id is not null) select queue_id as id from t where queue_id is not null [42102-191]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.command.Parser.readTableOrView(Parser.java:5349)
	at org.h2.command.Parser.readTableFilter(Parser.java:1245)
	at org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:1884)
	at org.h2.command.Parser.parseSelectSimple(Parser.java:2032)
	at org.h2.command.Parser.parseSelectSub(Parser.java:1878)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:1699)
	at org.h2.command.Parser.parseSelectSub(Parser.java:1874)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:1699)
	at org.h2.command.Parser.parseSelect(Parser.java:1687)
	at org.h2.command.Parser.parseWith(Parser.java:4745)
	at org.h2.command.Parser.parsePrepared(Parser.java:479)
	at org.h2.command.Parser.parse(Parser.java:315)
	at org.h2.command.Parser.parse(Parser.java:287)
	at org.h2.command.Parser.prepareCommand(Parser.java:252)
	at org.h2.engine.Session.prepareLocal(Session.java:560)
	at org.h2.engine.Session.prepareCommand(Session.java:501)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1188)
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:73)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:276)
	at org.skife.jdbi.v2.DefaultStatementBuilder.create(DefaultStatementBuilder.java:54)
	at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1302)
...
```

--- a/digdag-cli/src/main/java/io/digdag/cli/Check.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/Check.java
@@ -1,6 +1,7 @@
 package io.digdag.cli;
 
 import java.io.PrintStream;
+import java.io.IOException;
 import java.util.Map;
 import java.util.Set;
 import java.util.HashMap;
@@ -90,18 +91,24 @@ public class Check
         return systemExit(error);
     }
 
-    public void check(String workflowNameArg) throws Exception
+    public void check(String workflowNameArg)
+            throws Exception
     {
-        Injector injector = new DigdagEmbed.Bootstrap()
-            .withWorkflowExecutor(false)
-            .withScheduleExecutor(false)
-            .withLocalAgent(false)
-            .addModules(binder -> {
-                binder.bind(YamlMapper.class).in(Scopes.SINGLETON);
-            })
-            .initialize()
-            .getInjector();
+        try (DigdagEmbed digdag = new DigdagEmbed.Bootstrap()
+                .withWorkflowExecutor(false)
+                .withScheduleExecutor(false)
+                .withLocalAgent(false)
+                .addModules(binder -> {
+                    binder.bind(YamlMapper.class).in(Scopes.SINGLETON);
+                })
+                .initializeWithoutShutdownHook()) {
+            check(digdag.getInjector(), workflowNameArg);
+        }
+    }
 
+    private void check(Injector injector, String workflowNameArg)
+            throws IOException
+    {
         final ConfigFactory cf = injector.getInstance(ConfigFactory.class);
         final ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);
         final ProjectArchiveLoader projectLoader = injector.getInstance(ProjectArchiveLoader.class);

--- a/digdag-cli/src/main/java/io/digdag/cli/Run.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/Run.java
@@ -47,6 +47,7 @@ import io.digdag.core.archive.ProjectArchiveLoader;
 import io.digdag.core.repository.StoredRevision;
 import io.digdag.core.repository.StoredWorkflowDefinition;
 import io.digdag.core.repository.ResourceNotFoundException;
+import io.digdag.core.repository.ResourceConflictException;
 import io.digdag.core.session.ArchivedTask;
 import io.digdag.core.session.StoredSessionAttemptWithSession;
 import io.digdag.core.session.TaskStateCode;
@@ -210,23 +211,29 @@ public class Run
 
     private static final List<Long> USE_ALL = null;
 
-    public void run(String workflowNameArg, String matchPattern) throws Exception
+    public void run(String workflowNameArg, String matchPattern)
+            throws Exception
     {
         Properties systemProps = loadSystemProperties();
 
-        Injector injector = new DigdagEmbed.Bootstrap()
-            .setSystemPlugins(loadSystemPlugins(systemProps))
-            .addModules(binder -> {
-                binder.bind(ResumeStateManager.class).in(Scopes.SINGLETON);
-                binder.bind(YamlMapper.class).in(Scopes.SINGLETON);  // used by ResumeStateManager
-                binder.bind(Run.class).toInstance(this);  // used by OperatorManagerWithSkip
-            })
-            .overrideModulesWith((binder) -> {
-                binder.bind(OperatorManager.class).to(OperatorManagerWithSkip.class).in(Scopes.SINGLETON);
-            })
-            .initialize()
-            .getInjector();
+        try (DigdagEmbed digdag = new DigdagEmbed.Bootstrap()
+                .setSystemPlugins(loadSystemPlugins(systemProps))
+                .addModules(binder -> {
+                    binder.bind(ResumeStateManager.class).in(Scopes.SINGLETON);
+                    binder.bind(YamlMapper.class).in(Scopes.SINGLETON);  // used by ResumeStateManager
+                    binder.bind(Run.class).toInstance(this);  // used by OperatorManagerWithSkip
+                })
+                .overrideModulesWith((binder) -> {
+                    binder.bind(OperatorManager.class).to(OperatorManagerWithSkip.class).in(Scopes.SINGLETON);
+                })
+                .initializeWithoutShutdownHook()) {
+            run(systemProps, digdag.getInjector(), workflowNameArg, matchPattern);
+        }
+    }
 
+    private void run(Properties systemProps, Injector injector, String workflowNameArg, String matchPattern)
+        throws IOException, TaskMatchPattern.MultipleTaskMatchException, TaskMatchPattern.NoMatchException, ResourceNotFoundException, ResourceConflictException, SystemExitException, InterruptedException
+    {
         final LocalSite localSite = injector.getInstance(LocalSite.class);
         final ConfigFactory cf = injector.getInstance(ConfigFactory.class);
         final ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);

--- a/digdag-cli/src/main/java/io/digdag/cli/Sched.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/Sched.java
@@ -81,28 +81,31 @@ public class Sched
     }
 
     private void sched()
-            throws ServletException, IOException
+            throws ServletException, Exception
     {
         // use memory database by default
         if (database == null) {
             memoryDatabase = true;
         }
 
-        Injector injector = new DigdagEmbed.Bootstrap()
-            .withWorkflowExecutor(false)
-            .withScheduleExecutor(false)
-            .withLocalAgent(false)
-            .initialize()
-            .getInjector();
+        Properties props;
 
-        ConfigFactory cf = injector.getInstance(ConfigFactory.class);
-        ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);
+        try (DigdagEmbed digdag = new DigdagEmbed.Bootstrap()
+                .withWorkflowExecutor(false)
+                .withScheduleExecutor(false)
+                .withLocalAgent(false)
+                .initializeWithoutShutdownHook()) {
+            Injector injector = digdag.getInjector();
 
-        Config overwriteParams = loadParams(cf, loader, loadSystemProperties(), paramsFile, params);
+            ConfigFactory cf = injector.getInstance(ConfigFactory.class);
+            ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);
 
-        Properties props = buildServerProperties();
-        props.setProperty(SYSTEM_CONFIG_AUTO_LOAD_LOCAL_PROJECT_KEY, projectDirName != null ? projectDirName : "");  // Properties can't store null
-        props.setProperty(SYSTEM_CONFIG_LOCAL_OVERWRITE_PARAMS, overwriteParams.toString());
+            Config overwriteParams = loadParams(cf, loader, loadSystemProperties(), paramsFile, params);
+
+            props = buildServerProperties();
+            props.setProperty(SYSTEM_CONFIG_AUTO_LOAD_LOCAL_PROJECT_KEY, projectDirName != null ? projectDirName : "");  // Properties can't store null
+            props.setProperty(SYSTEM_CONFIG_LOCAL_OVERWRITE_PARAMS, overwriteParams.toString());
+        }
 
         ServerBootstrap.startServer(localVersion, props, SchedulerServerBootStrap.class);
     }

--- a/digdag-cli/src/main/java/io/digdag/cli/Show.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/Show.java
@@ -51,13 +51,18 @@ public class Show
     private void show(String workflowPath)
             throws Exception
     {
-        Injector injector = new DigdagEmbed.Bootstrap()
-            .withWorkflowExecutor(false)
-            .withScheduleExecutor(false)
-            .withLocalAgent(false)
-            .initialize()
-            .getInjector();
+        try (DigdagEmbed digdag = new DigdagEmbed.Bootstrap()
+                .withWorkflowExecutor(false)
+                .withScheduleExecutor(false)
+                .withLocalAgent(false)
+                .initialize()) {
+            show(digdag.getInjector(), workflowPath);
+        }
+    }
 
+    private void show(Injector injector, String workflowPath)
+            throws Exception
+    {
         final ConfigFactory cf = injector.getInstance(ConfigFactory.class);
         final ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);
         final WorkflowCompiler compiler = injector.getInstance(WorkflowCompiler.class);

--- a/digdag-cli/src/main/java/io/digdag/cli/client/Archive.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/client/Archive.java
@@ -73,9 +73,9 @@ public class Archive
     }
 
     private void archive()
-            throws IOException
+            throws Exception
     {
-        Injector injector = new DigdagEmbed.Bootstrap()
+        try (DigdagEmbed digdag = new DigdagEmbed.Bootstrap()
                 .withWorkflowExecutor(false)
                 .withScheduleExecutor(false)
                 .withLocalAgent(false)
@@ -85,9 +85,14 @@ public class Archive
                     binder.bind(PrintStream.class).annotatedWith(StdOut.class).toInstance(out);
                     binder.bind(PrintStream.class).annotatedWith(StdErr.class).toInstance(err);
                 })
-                .initialize()
-                .getInjector();
+                .initializeWithoutShutdownHook()) {
+            archive(digdag.getInjector());
+        }
+    }
 
+    private void archive(Injector injector)
+            throws IOException
+    {
         ConfigFactory cf = injector.getInstance(ConfigFactory.class);
         ConfigLoaderManager loader = injector.getInstance(ConfigLoaderManager.class);
 

--- a/digdag-cli/src/main/java/io/digdag/cli/client/Delete.java
+++ b/digdag-cli/src/main/java/io/digdag/cli/client/Delete.java
@@ -16,7 +16,6 @@ import io.digdag.cli.TimeUtil;
 import io.digdag.client.DigdagClient;
 import io.digdag.client.api.RestProject;
 import io.digdag.client.config.Config;
-import io.digdag.core.DigdagEmbed;
 import io.digdag.core.Version;
 
 import static io.digdag.cli.SystemExitException.systemExit;

--- a/digdag-core/src/main/java/io/digdag/core/DigdagEmbed.java
+++ b/digdag-core/src/main/java/io/digdag/core/DigdagEmbed.java
@@ -43,6 +43,7 @@ import io.digdag.client.config.ConfigFactory;
 import io.digdag.client.api.JacksonTimeModule;
 
 public class DigdagEmbed
+        implements AutoCloseable
 {
     public static class Bootstrap
     {
@@ -122,7 +123,7 @@ public class DigdagEmbed
             return build(true);
         }
 
-        public DigdagEmbed initializeCloseable()
+        public DigdagEmbed initializeWithoutShutdownHook()
         {
             return build(false);
         }
@@ -254,7 +255,8 @@ public class DigdagEmbed
         return getInjector().getInstance(LocalSite.class);
     }
 
-    public void destroy() throws Exception
+    @Override
+    public void close() throws Exception
     {
         injector.destroy();
     }

--- a/digdag-core/src/main/java/io/digdag/core/agent/LocalAgentManager.java
+++ b/digdag-core/src/main/java/io/digdag/core/agent/LocalAgentManager.java
@@ -1,19 +1,19 @@
 package io.digdag.core.agent;
 
+import java.util.function.Supplier;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ExecutorService;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
 import com.google.inject.Inject;
-import com.google.common.base.*;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import io.digdag.core.queue.TaskQueueManager;
 
 public class LocalAgentManager
 {
-    private final AgentConfig config;
-    private final AgentId agentId;
-    private final TaskQueueManager queueManager;
-    private final OperatorManager operatorManager;
-    private final ExecutorService executor;
+    private final Supplier<MultiThreadAgent> agentFactory;
+    private Thread thread;
+    private MultiThreadAgent agent;
 
     @Inject
     public LocalAgentManager(
@@ -22,35 +22,36 @@ public class LocalAgentManager
             TaskQueueManager queueManager,
             OperatorManager operatorManager)
     {
-        this.config = config;
-        this.agentId = agentId;
-        this.queueManager = queueManager;
-        this.operatorManager = operatorManager;
         if (config.getEnabled()) {
-            this.executor = Executors.newCachedThreadPool(
-                    new ThreadFactoryBuilder()
-                    .setDaemon(true)
-                    .setNameFormat("local-agent-%d")
-                    .build());
+            this.agentFactory = () -> new MultiThreadAgent(config, agentId, queueManager.getInProcessTaskQueueClient(), operatorManager);
         }
         else {
-            this.executor = null;
+            this.agentFactory = null;
         }
     }
 
-    // TODO stop LocalAgent at @PreDestroy
-
+    @PostConstruct
     public void start()
     {
-        if (executor != null) {
-            executor.submit(
-                    new LocalAgent(
-                        config,
-                        agentId,
-                        queueManager.getInProcessTaskQueueClient(),
-                        operatorManager
-                    )
-                );
+        if (agentFactory != null) {
+            agent = agentFactory.get();
+            Thread thread = new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("local-agent-%d")
+                .build()
+                .newThread(agent);
+            thread.start();
+            this.thread = thread;
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+        throws InterruptedException
+    {
+        if (thread != null) {
+            agent.shutdown();
+            thread.join();
         }
     }
 }

--- a/digdag-core/src/main/java/io/digdag/core/agent/LocalAgentModule.java
+++ b/digdag-core/src/main/java/io/digdag/core/agent/LocalAgentModule.java
@@ -16,7 +16,6 @@ public class LocalAgentModule
     @Override
     public void configure(Binder binder)
     {
-        binder.bind(LocalAgentManager.class).in(Scopes.SINGLETON);
         binder.bind(AgentConfig.class).toProvider(AgentConfigProvider.class).in(Scopes.SINGLETON);
         binder.bind(WorkspaceManager.class).to(NoopWorkspaceManager.class).in(Scopes.SINGLETON);
         binder.bind(TaskCallbackApi.class).to(InProcessTaskCallbackApi.class).in(Scopes.SINGLETON);
@@ -27,24 +26,6 @@ public class LocalAgentModule
         taskExecutorBinder.addBinding().to(RequireOperatorFactory.class).in(Scopes.SINGLETON);
         taskExecutorBinder.addBinding().to(CallOperatorFactory.class).in(Scopes.SINGLETON);
 
-        binder.bind(LocalAgentStarter.class).asEagerSingleton();
-    }
-
-    public static class LocalAgentStarter
-    {
-        private final LocalAgentManager localAgentManager;
-
-        @Inject
-        public LocalAgentStarter(
-                LocalAgentManager localAgentManager)
-        {
-            this.localAgentManager = localAgentManager;
-        }
-
-        @PostConstruct
-        public void start()
-        {
-            localAgentManager.start();
-        }
+        binder.bind(LocalAgentManager.class).asEagerSingleton();
     }
 }

--- a/digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java
+++ b/digdag-core/src/main/java/io/digdag/core/agent/MultiThreadAgent.java
@@ -3,17 +3,18 @@ package io.digdag.core.agent;
 import java.util.List;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ExecutorService;
-import com.google.common.base.*;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import io.digdag.spi.TaskRequest;
 import io.digdag.spi.TaskQueueClient;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class LocalAgent
+public class MultiThreadAgent
         implements Runnable
 {
-    private static final Logger logger = LoggerFactory.getLogger(LocalAgent.class);
+    private static final Logger logger = LoggerFactory.getLogger(MultiThreadAgent.class);
 
     private final AgentConfig config;
     private final AgentId agentId;
@@ -22,39 +23,30 @@ public class LocalAgent
     private final ExecutorService executor;
     private volatile boolean stop = false;
 
-    public LocalAgent(AgentConfig config, AgentId agentId,
+    public MultiThreadAgent(AgentConfig config, AgentId agentId,
             TaskQueueClient queue, OperatorManager runner)
     {
         this.agentId = agentId;
         this.config = config;
         this.queue = queue;
         this.runner = runner;
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setDaemon(false)  // make them non-daemon threads so that shutting down agent doesn't kill operator execution
+            .setNameFormat("task-thread-%d")
+            .build();
         if (config.getMaxThreads() > 0) {
-            this.executor = Executors.newFixedThreadPool(
-                    config.getMaxThreads(),
-                    new ThreadFactoryBuilder()
-                    .setDaemon(true)
-                    .setNameFormat("task-thread-%d")
-                    .build());
+            this.executor = Executors.newFixedThreadPool(config.getMaxThreads(), threadFactory);
         }
         else {
-            this.executor = Executors.newCachedThreadPool(
-                    new ThreadFactoryBuilder()
-                    .setDaemon(true)
-                    .setNameFormat("task-thread-%d")
-                    .build());
+            this.executor = Executors.newCachedThreadPool(threadFactory);
         }
-    }
-
-    public void stop()
-    {
-        stop = true;
     }
 
     public void shutdown()
     {
+        stop = true;
+        queue.interruptLocalWait();
         executor.shutdown();
-        // TODO wait for shutdown completion?
     }
 
     @Override
@@ -62,7 +54,6 @@ public class LocalAgent
     {
         while (!stop) {
             try {
-                // TODO implement task heartbeat that calls queue.taskHeartbeat using a background thread
                 List<TaskRequest> reqs = queue.lockSharedTasks(3, agentId.toString(), config.getLockRetentionTime(), 1000);
                 for (TaskRequest req : reqs) {
                     executor.submit(() -> {

--- a/digdag-core/src/main/java/io/digdag/core/database/DatabaseTaskQueueFactory.java
+++ b/digdag-core/src/main/java/io/digdag/core/database/DatabaseTaskQueueFactory.java
@@ -81,7 +81,7 @@ public class DatabaseTaskQueueFactory
         }
 
         @Override
-        public List<TaskRequest> lockSharedTasks(int limit, String agentId, int lockSeconds, long maxSleepMillis)
+        public List<TaskRequest> lockSharedTasks(int limit, String agentId, int lockSeconds, long maxWaitMillis)
         {
             ImmutableList.Builder<TaskRequest> builder = ImmutableList.builder();
             for (LockResult lock : store.lockSharedTasks(limit, agentId, lockSeconds)) {
@@ -94,17 +94,17 @@ public class DatabaseTaskQueueFactory
                 }
             }
             List<TaskRequest> result = builder.build();
-            if (result.isEmpty() && maxSleepMillis >= 0) {
-                sleepUntilEnqueue(sharedTaskSleepHelper, maxSleepMillis);
+            if (result.isEmpty() && maxWaitMillis >= 0) {
+                sleepUntilEnqueue(sharedTaskSleepHelper, maxWaitMillis);
             }
             return result;
         }
 
-        private void sleepUntilEnqueue(Object helper, long maxSleepMillis)
+        private void sleepUntilEnqueue(Object helper, long maxWaitMillis)
         {
             synchronized (helper) {
                 try {
-                    helper.wait(maxSleepMillis);
+                    helper.wait(maxWaitMillis);
                 }
                 catch (InterruptedException ex) {
                     return;
@@ -118,6 +118,13 @@ public class DatabaseTaskQueueFactory
             synchronized (helper) {
                 helper.notifyAll();
             }
+        }
+
+        @Override
+        public void interruptLocalWait()
+        {
+            noticeEnqueue(sharedTaskSleepHelper);
+            noticeEnqueue(taskSleepHelper);
         }
 
         @Override

--- a/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowCompilerTest.java
+++ b/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowCompilerTest.java
@@ -8,6 +8,7 @@ import io.digdag.client.config.ConfigFactory;
 import io.digdag.core.DigdagEmbed;
 import io.digdag.core.config.YamlConfigLoader;
 import org.junit.Before;
+import org.junit.After;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;

--- a/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowCompilerTest.java
+++ b/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowCompilerTest.java
@@ -33,6 +33,13 @@ public class WorkflowCompilerTest
         compiler = new WorkflowCompiler();
     }
 
+    @After
+    public void destroy()
+        throws Exception
+    {
+        embed.close();
+    }
+
     private Config loadYamlResource(String name)
     {
         try {

--- a/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowExecutorCasesTest.java
+++ b/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowExecutorCasesTest.java
@@ -57,7 +57,7 @@ public class WorkflowExecutorCasesTest
     public void destroy()
         throws Exception
     {
-        embed.destroy();
+        embed.close();
     }
 
     @Test

--- a/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowTestingUtils.java
+++ b/digdag-core/src/test/java/io/digdag/core/workflow/WorkflowTestingUtils.java
@@ -32,7 +32,7 @@ public class WorkflowTestingUtils
             .overrideModulesWith((binder) -> {
                 binder.bind(DatabaseConfig.class).toInstance(getEnvironmentDatabaseConfig());
             })
-            .initializeCloseable();
+            .initializeWithoutShutdownHook();
         cleanDatabase(embed);
         return embed;
     }

--- a/digdag-server/src/main/java/io/digdag/server/ServerBootstrap.java
+++ b/digdag-server/src/main/java/io/digdag/server/ServerBootstrap.java
@@ -84,9 +84,12 @@ public class ServerBootstrap
 
         Version version = Version.of(context.getInitParameter(VERSION_INIT_PARAMETER_KEY));
 
-        Injector injector = bootstrap(new DigdagEmbed.Bootstrap(), serverConfig, version)
-            .initialize()
-            .getInjector();
+        // GuiceRsServletContainerInitializer closes Digdag (through Injector implementing AutoCloseable)
+        // when servlet context is destroyed.
+        DigdagEmbed digdag = bootstrap(new DigdagEmbed.Bootstrap(), serverConfig, version)
+            .initialize();
+
+        Injector injector = digdag.getInjector();
 
         if (serverConfig.getExecutorEnabled()) {
             // TODO create global site

--- a/digdag-spi/src/main/java/io/digdag/spi/TaskQueueClient.java
+++ b/digdag-spi/src/main/java/io/digdag/spi/TaskQueueClient.java
@@ -4,7 +4,10 @@ import java.util.List;
 
 public interface TaskQueueClient
 {
-    List<TaskRequest> lockSharedTasks(int limit, String agentId, int lockSeconds, long maxSleepMillis);
+    List<TaskRequest> lockSharedTasks(int limit, String agentId, int lockSeconds, long maxWaitMillis);
+
+    default void interruptLocalWait()
+    { }
 
     // TODO lockTasks (of custom queue) is not implemented yet
 

